src/provider/hyrax_pc.rs

```diff
@@ -0,0 +1,414 @@
+      self.clone()
+    } else {
+      let result = self
+        .comm
+        .iter()
+        .zip_longest(other.comm.iter())
+        .map(|x| match x {
+          Both(a, b) => a + b,
+          Left(a) => *a,
+          Right(b) => *b,
+        })
+        .collect();
+      HyraxCommitment {
+        comm: result,
+        is_default: self.is_default,
+      }
+    }
+  }
+}
+
+macro_rules! define_add_variants {
+  (G = $g:path, LHS = $lhs:ty, RHS = $rhs:ty, Output = $out:ty) => {
+    impl<'b, G: $g> Add<&'b $rhs> for $lhs {
+      type Output = $out;
+      fn add(self, rhs: &'b $rhs) -> $out {
+        &self + rhs
+      }
+    }
+
+    impl<'a, G: $g> Add<$rhs> for &'a $lhs {
+      type Output = $out;
+      fn add(self, rhs: $rhs) -> $out {
+        self + &rhs
+      }
+    }
+
+    impl<G: $g> Add<$rhs> for $lhs {
+      type Output = $out;
+      fn add(self, rhs: $rhs) -> $out {
+        &self + &rhs
+      }
+    }
+  };
+}
+
+macro_rules! define_add_assign_variants {
+  (G = $g:path, LHS = $lhs:ty, RHS = $rhs:ty) => {
+    impl<G: $g> AddAssign<$rhs> for $lhs {
+      fn add_assign(&mut self, rhs: $rhs) {
+        *self += &rhs;
+      }
+    }
+  };
+}
+
+define_add_assign_variants!(G = Group, LHS = HyraxCommitment<G>, RHS = HyraxCommitment<G>);
+define_add_variants!(G = Group, LHS = HyraxCommitment<G>, RHS = HyraxCommitment<G>, Output = HyraxCommitment<G>);
+
+/// Provides a commitment engine
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct HyraxCommitmentEngine<G: Group> {
+  _p: PhantomData<G>,
+}
+
+impl<G: Group> CommitmentEngineTrait<G> for HyraxCommitmentEngine<G> {
+  type CommitmentKey = HyraxCommitmentKey<G>;
+  type Commitment = HyraxCommitment<G>;
+
+  /// Derives generators for Hyrax PC, where num_vars is the number of variables in multilinear poly
+  fn setup(label: &'static [u8], n: usize) -> Self::CommitmentKey {
+    let num_vars = n.next_power_of_two().log_2();
+    let (_left, right) = EqPolynomial::<G::Scalar>::compute_factored_lens(num_vars);
+    let ck = PedersenCommitmentEngine::setup(label, (2usize).pow(right as u32));
+    HyraxCommitmentKey {
+      ck,
+      _p: Default::default(),
+    }
+  }
+
+  fn commit(ck: &Self::CommitmentKey, v: &[G::Scalar]) -> Self::Commitment {
+    let poly = MultilinearPolynomial::new(v.to_vec());
+    let n = poly.len();
+    let ell = poly.get_num_vars();
+    assert_eq!(n, (2usize).pow(ell as u32));
+
+    let (left_num_vars, right_num_vars) = EqPolynomial::<G::Scalar>::compute_factored_lens(ell);
+    let L_size = (2usize).pow(left_num_vars as u32);
+    let R_size = (2usize).pow(right_num_vars as u32);
+    assert_eq!(L_size * R_size, n);
+
+    let comm = (0..L_size)
+      .collect::<Vec<usize>>()
+      .into_par_iter()
+      .map(|i| {
+        PedersenCommitmentEngine::commit(&ck.ck, &poly.get_Z()[R_size * i..R_size * (i + 1)])
+      })
+      .collect();
+
+    HyraxCommitment {
+      comm,
+      is_default: false,
```

Comment:
I see. Rather than a boolean flag is_default, I think what you want to employ in the `HyraxCommitment` is a [`std::cell::OnceCell`](https://doc.rust-lang.org/stable/std/cell/struct.OnceCell.html).
This `OnceCell` represents a slot that you can fill "only once" (with any value you want), after which it becomes immutable. This should allow you to implement logic for commitment and operations thereof without having to rely on error management, or flag-checking logic.